### ISSUE TITLE
[bug] conan_server authorization oddity

### ISSUE THREAD

**Issue by hlewin** [conan-io/conan#11185]
I noticed a quite unexpected and potentially dangerous oddity with the way `conan_server` handles authorization.

When a package without user and channel is uploaded like "MyPackage/1.0@" the reference gets stored with user=None.
What is odd is that `conan_server` does not check any permission-rules when the user requesting data equals the user of the ref.
This means that even if the permission rules explicitly have a rule like
```
[read_permissions]
*/*@*/*: some_user
```
an anonymous user can still read the package. Further on anonymous users can always read the package-index (`conan search...`) as those seem to be handled as ref belonging to user `None`. On the other hand, if you are logged in as some other user "some_user2", you will be denied to read the package.

This behaviour is due to a shortcut in `authorize.py` (in conans/server/service)  which reads
```
class BasicAuthorizer(Authorizer):
    """
    Reads permissions from the config file (server.cfg)
    """
....
    def check_read_conan(self, username, ref):
        """
        username: User that request to read the conans
        ref: ConanFileReference
        """
        if ref.user == username:
            return

        self._check_any_rule_ok(username, self.read_permissions, ref)
```
Here you can clearly see that rules are not checked for such cases.

This is quite unexpected, not clearly documented and this potentially dangerous.
Anonymous users will always be able to search server-contents and download packages without user/channel - no matter what the server.conf says.

**Comment by memsharded** [comment]
Hi @hlewin 

Yes, this looks like it could be an issue. As packages without user/channel were introduced later, this was probably not completely considered in conan_server.

While this is investigated, these would be some recommendations:
- In general, user packages should have a ``user/channel``. Only packages from ConanCenter don't have user/channel, and those are public.
- You can use an ArtifactoryCE. It is completely free, easy to run, it decouples permission from the ``user/channel`` user name, it has a web UI, and it can scale much better than ``conan_server``.

**Comment by joshvazquez** [comment]
I've encountered this as well while writing a custom Authorizer. Is it also recommended to specify user and channel for packages shared within a company? What if the owner should be a team and not a specific person?

ArtifactoryCE is not suitable for me because I need custom authentication & authorization.

It's difficult to trace the calls because of how Bottle works, could you direct me to where API calls are made with user None before authenticating?

**Comment by joshvazquez** [comment]
It would be smart if I did the trace on the client instead of the server...

Anyway it seems to be [here](https://github.com/conan-io/conan/blob/develop/conans/client/rest/auth_manager.py#L40) that `get_recipe_manifest` is the first thing called without authentication. I'm going to experiment with this file and see if I can make it always auth.

**Comment by joshvazquez** [comment]
Adding `self._auth_manager.call_rest_api_method(remote, "check_credentials")` above https://github.com/conan-io/conan/blob/develop/conans/client/remote_manager.py#L290 seems to work.

**Comment by memsharded** [comment]
> Is it also recommended to specify user and channel for packages shared within a company? What if the owner should be a team and not a specific person?

Yes, it is good practice to setup user/channel (in 2.0 only user is possible), for your own packages. Typically named after the company or team, depending on the scope, and occasionally after individual developers just while playing around with new dependencies and probably only uploading to "playground" repositories, not to develop/production

> Adding self._auth_manager.call_rest_api_method(remote, "check_credentials") above https://github.com/conan-io/conan/blob/develop/conans/client/remote_manager.py#L290 seems to work.

This would introduce an extra duplicated auth API call before every single API requests, doubling communication times, this doesn't sound possible. It seems that this should be addressed in the server side.

**Comment by joshvazquez** [comment]
You are right, this is at best a hack. But for reasons I can't explain, this is working now even without that change.

My auth system provides the username in a header. Before, my Authorizer was receiving None user during the call to `get_recipe_manifest_url`, raising AuthenticationException, and prompting the client for credentials. Now, it is correctly receiving the username in the header and not raising an exception. And username/password auth via the HTTP Basic and JWT plugins still works.

**Comment by memsharded** [comment]
Sorry this was not followed up back then.

I have opened https://github.com/conan-io/conan/pull/17132 that will close this ticket for next Conan 2.9

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/17132 was merged, it will be in next Conan 2.9
